File: lib/block-cipher.ts
export const BLOCK_SIZE = 16;

/**
 * Konversi string Teks ke Uint8Array menggunakan UTF-8 encoding
 */
export function stringToBytes(str: string): Uint8Array {
  return new TextEncoder().encode(str);
}

/**
 * Konversi Uint8Array kembali ke string Teks
 */
export function bytesToString(bytes: Uint8Array): string {
  return new TextDecoder().decode(bytes);
}

/**
 * Padding sebuah Key string menjadi tepat 16 byte.
 * Jika kurang, diulang. Jika lebih dipotong.
 */
export function padKey(keyStr: string): Uint8Array {
  const bytes = stringToBytes(keyStr);
  const key = new Uint8Array(BLOCK_SIZE);
  if (bytes.length === 0) {
    // default key if empty
    key.fill(0xAA);
    return key;
  }
  for (let i = 0; i < BLOCK_SIZE; i++) {
    key[i] = bytes[i % bytes.length];
  }
  return key;
}

/**
 * PKCS#7 Padding
 */
export function pkcs7Pad(data: Uint8Array): Uint8Array {
  const padLen = BLOCK_SIZE - (data.length % BLOCK_SIZE);
  const padded = new Uint8Array(data.length + padLen);
  padded.set(data);
  for (let i = data.length; i < padded.length; i++) {
    padded[i] = padLen;
  }
  return padded;
}

export function pkcs7Unpad(data: Uint8Array): Uint8Array {
  if (data.length === 0) return data;
  const padLen = data[data.length - 1];
  if (padLen > 0 && padLen <= BLOCK_SIZE) {
    return data.slice(0, data.length - padLen);
  }
  return data; // If invalid padding, return as is safely
}

// ==========================================
//   Electronic Codebook (ECB) Mode
// ==========================================

function encryptBlockECB(block: Uint8Array, key: Uint8Array): Uint8Array {
  // 1. Substitusi (Add & Modulo)
  const subOut = new Uint8Array(BLOCK_SIZE);
  for (let i = 0; i < BLOCK_SIZE; i++) {
    subOut[i] = (block[i] + key[i]) % 256;
  }
  
  // 2. Rotasi (Circular Shift Left array elements by 3)
  const rotOut = new Uint8Array(BLOCK_SIZE);
  const shift = 3;
  for (let i = 0; i < BLOCK_SIZE; i++) {
    rotOut[i] = subOut[(i + shift) % BLOCK_SIZE];
  }
  
  // 3. Transposisi (Swap pattern)
  // Membaca seperti kolom per kolom matriks 4x4
  const transposePattern = [0, 4, 8, 12, 1, 5, 9, 13, 2, 6, 10, 14, 3, 7, 11, 15];
  const finalOut = new Uint8Array(BLOCK_SIZE);
  for (let i = 0; i < BLOCK_SIZE; i++) {
    finalOut[i] = rotOut[transposePattern[i]];
  }
  
  return finalOut;
}

function decryptBlockECB(block: Uint8Array, key: Uint8Array): Uint8Array {
  // 3. Undo Transposisi
  const transposePattern = [0, 4, 8, 12, 1, 5, 9, 13, 2, 6, 10, 14, 3, 7, 11, 15];
  const rotOut = new Uint8Array(BLOCK_SIZE);
  for (let i = 0; i < BLOCK_SIZE; i++) {
    rotOut[transposePattern[i]] = block[i];
  }
  
  // 2. Undo Rotasi (Circular Shift Right by 3)
  const shift = 3;
  const subOut = new Uint8Array(BLOCK_SIZE);
  for (let i = 0; i < BLOCK_SIZE; i++) {
    // if subOut is rotated left to rotOut, then reading subOut:
    subOut[(i + shift) % BLOCK_SIZE] = rotOut[i];
  }
  
  // 1. Undo Substitusi
  const finalOut = new Uint8Array(BLOCK_SIZE);
  for (let i = 0; i < BLOCK_SIZE; i++) {
    finalOut[i] = (subOut[i] - key[i] + 256) % 256;
  }
  
  return finalOut;
}

export function encryptECB(data: Uint8Array, keyStr: string, usePadding: boolean): Uint8Array {
  const key = padKey(keyStr);
  const processData = usePadding ? pkcs7Pad(data) : data;
  const numBlocks = Math.floor(processData.length / BLOCK_SIZE);
  const finalSize = usePadding ? processData.length : processData.length;
  const out = new Uint8Array(finalSize);
  out.set(processData); // Copy initial state (useful for tailing bytes without padding)

  for (let i = 0; i < numBlocks; i++) {
    const offset = i * BLOCK_SIZE;
    const block = processData.slice(offset, offset + BLOCK_SIZE);
    const cipherBlock = encryptBlockECB(block, key);
    out.set(cipherBlock, offset);
  }
  return out;
}

export function decryptECB(data: Uint8Array, keyStr: string, usePadding: boolean): Uint8Array {
  const key = padKey(keyStr);
  const numBlocks = Math.floor(data.length / BLOCK_SIZE);
  const out = new Uint8Array(data.length);
  out.set(data); // preserve tails

  for (let i = 0; i < numBlocks; i++) {
    const offset = i * BLOCK_SIZE;
    const block = data.slice(offset, offset + BLOCK_SIZE);
    const plainBlock = decryptBlockECB(block, key);
    out.set(plainBlock, offset);
  }
  return usePadding ? pkcs7Unpad(out) : out;
}

// ==========================================
//   Cipher Block Chaining (CBC) Mode
//   XOR-sub -> Rotate-Left-5 -> Pairwise-swap+NOT
// ==========================================

function encryptBlockCBC(block: Uint8Array, key: Uint8Array): Uint8Array {
  // 1. Substitusi (XOR dengan Key)
  const subOut = new Uint8Array(BLOCK_SIZE);
  for (let i = 0; i < BLOCK_SIZE; i++) {
    subOut[i] = block[i] ^ key[i];
  }

  // 2. Rotasi (Circular Shift Left array elements by 5)
  const rotOut = new Uint8Array(BLOCK_SIZE);
  const shift = 5;
  for (let i = 0; i < BLOCK_SIZE; i++) {
    rotOut[i] = subOut[(i + shift) % BLOCK_SIZE];
  }

  // 3. Permutasi (Pairwise Swap + Bitwise NOT)
  // [0] ditukar [1] dan di Invert, [2] ditukar [3] dan Invert, dst.
  const finalOut = new Uint8Array(BLOCK_SIZE);
  for (let i = 0; i < BLOCK_SIZE; i += 2) {
    finalOut[i]     = (~rotOut[i+1]) & 0xFF;
    finalOut[i+1]   = (~rotOut[i])   & 0xFF;
  }

  return finalOut;
}

function decryptBlockCBC(block: Uint8Array, key: Uint8Array): Uint8Array {
  // 3. Undo Permutasi (Pairwise Swap + Bitwise NOT bersifat simetris jika dipanggil kembali)
  const rotOut = new Uint8Array(BLOCK_SIZE);
  for (let i = 0; i < BLOCK_SIZE; i += 2) {
    rotOut[i]     = (~block[i+1]) & 0xFF;
    rotOut[i+1]   = (~block[i])   & 0xFF;
  }

  // 2. Undo Rotasi (Circular Shift Right by 5)
  const shift = 5;
  const subOut = new Uint8Array(BLOCK_SIZE);
  for (let i = 0; i < BLOCK_SIZE; i++) {
    subOut[(i + shift) % BLOCK_SIZE] = rotOut[i];
  }

  // 1. Undo Substitusi (XOR juga simetris)
  const finalOut = new Uint8Array(BLOCK_SIZE);
  for (let i = 0; i < BLOCK_SIZE; i++) {
    finalOut[i] = subOut[i] ^ key[i];
  }
  return finalOut;
}

function xorBlocks(a: Uint8Array, b: Uint8Array): Uint8Array {
  const out = new Uint8Array(BLOCK_SIZE);
  for (let i = 0; i < BLOCK_SIZE; i++) {
    out[i] = a[i] ^ b[i];
  }
  return out;
}

export function encryptCBC(data: Uint8Array, keyStr: string, ivStr: string, usePadding: boolean): Uint8Array {
  const key = padKey(keyStr);
  const iv = padKey(ivStr); // IV uses same derivation technique as Key
  
  const processData = usePadding ? pkcs7Pad(data) : data;
  const numBlocks = Math.floor(processData.length / BLOCK_SIZE);
  const finalSize = usePadding ? processData.length : processData.length;
  const out = new Uint8Array(finalSize);
  out.set(processData);

  let prevCipherBlock = iv;

  for (let i = 0; i < numBlocks; i++) {
    const offset = i * BLOCK_SIZE;
    const block = processData.slice(offset, offset + BLOCK_SIZE);
    
    // CBC Step 1: XOR plaintext block with previous ciphertext block (or IV)
    const chainedBlock = xorBlocks(block, prevCipherBlock);
    
    // CBC Step 2: Encrypt
    const cipherBlock = encryptBlockCBC(chainedBlock, key);
    out.set(cipherBlock, offset);
    
    // Update vector for next chain
    prevCipherBlock = cipherBlock;
  }
  return out;
}

export function decryptCBC(data: Uint8Array, keyStr: string, ivStr: string, usePadding: boolean): Uint8Array {
  const key = padKey(keyStr);
  const iv = padKey(ivStr);
  
  const numBlocks = Math.floor(data.length / BLOCK_SIZE);
  const out = new Uint8Array(data.length);
  out.set(data);

  let prevCipherBlock = iv;

  for (let i = 0; i < numBlocks; i++) {
    const offset = i * BLOCK_SIZE;
    const cipherBlock = data.slice(offset, offset + BLOCK_SIZE);
    
    // CBC Step 1: Decrypt
    const decryptedBlock = decryptBlockCBC(cipherBlock, key);
    
    // CBC Step 2: XOR with previous ciphertext block (or IV)
    const plainBlock = xorBlocks(decryptedBlock, prevCipherBlock);
    out.set(plainBlock, offset);
    
    // Update vector for next chain using ORIGINAL ciphertext block before decryption
    prevCipherBlock = cipherBlock;
  }
  return usePadding ? pkcs7Unpad(out) : out;
}

// Helpers for Hex viewing
export function bytesToHex(bytes: Uint8Array): string {
  return Array.from(bytes).map(b => b.toString(16).padStart(2, '0')).join('');
}

export function hexToBytes(hex: string): Uint8Array {
  const cleanHex = hex.replace(/[^0-9a-fA-F]/g, '');
  const bytes = new Uint8Array(cleanHex.length / 2);
  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = parseInt(cleanHex.substring(i * 2, i * 2 + 2), 16);
  }
  return bytes;
}

File: block_cipher.py
BLOCK_SIZE = 16


# ─────────────────────────────────────────────
#  Utility Helpers
# ─────────────────────────────────────────────

def pad_key(key_str: str) -> bytes:
    """
    pad_key fungsi untuk memastikan bahwa kunci yang digunakan
    memiliki panjang 16 byte, menambah kekurangan byte
    dengan mengulang karakter sevara cycle.
    """
    raw = key_str.encode("utf-8")
    if len(raw) == 0:
        return bytes([0xAA] * BLOCK_SIZE)
    return bytes(raw[i % len(raw)] for i in range(BLOCK_SIZE))


def block_pad(data: bytes) -> bytes:
    """
    block padding:
    Menambahkan padding blok: tambahkan N byte dengan nilai N,
    selalu menambahkan walau udah pas
    """
    pad_len = BLOCK_SIZE - (len(data) % BLOCK_SIZE)
    return data + bytes([pad_len] * pad_len)


def block_unpad(data: bytes) -> bytes:
    """
    Menghapus padding: baca byte terakhir untuk mendapatkan panjang padding,
    validasi bahwa panjangnya 1-16, lalu hapus padding.
    """
    if not data:
        return data
    pad_len = data[-1]
    if 0 < pad_len <= BLOCK_SIZE:
        return data[:-pad_len]
    return data


def bytes_to_hex(data: bytes) -> str:
    return data.hex().upper()


def hex_to_bytes(hex_str: str) -> bytes:
    #Membersihkan dan mengubah hex menjadi byte
    clean = "".join(c for c in hex_str if c in "0123456789abcdefABCDEF")
    return bytes.fromhex(clean)


def xor_blocks(a: bytes, b: bytes) -> bytes:
    """Operasi XOR antara dua blok byte, menghasilkan blok baru dengan XOR byte-per-byte."""
    return bytes(x ^ y for x, y in zip(a, b))


# ─────────────────────────────────────────────
#  ECB - Electronic Codebook
#  Substitution -> Rotation -> Transposition
# ─────────────────────────────────────────────

TRANSPOSE_PATTERN = [0, 4, 8, 12, 
                     1, 5, 9, 13, 
                     2, 6, 10, 14, 
                     3, 7, 11, 15] # matriks 4x4 sebelumnya [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]

def _encrypt_block_ecb(block: bytes, key: bytes) -> bytes:
    block = bytearray(block)
    key   = bytearray(key)

    # Step 1 - Substitution: byte + key_byte (mod 256)
    sub = bytearray(BLOCK_SIZE)
    for i in range(BLOCK_SIZE):
        sub[i] = (block[i] + key[i]) % 256

    # Step 2 - Rotasi: Geser ke kiri secara sirkular sebanyak 3 posisi
    shift = 3
    rot = bytearray(BLOCK_SIZE)
    for i in range(BLOCK_SIZE):
        rot[i] = sub[(i + shift) % BLOCK_SIZE]

    # Step 3 - Transposition: memmetakan byte ke posisi baru berdasarkan transpose pattern (matrix 4x4)
    out = bytearray(BLOCK_SIZE)
    for i in range(BLOCK_SIZE):
        out[i] = rot[TRANSPOSE_PATTERN[i]]

    return bytes(out)


def _decrypt_block_ecb(block: bytes, key: bytes) -> bytes:
    block = bytearray(block)
    key   = bytearray(key)

    # Undo Step 3 - Membalikkan transposisi: tempatkan byte kembali ke posisi aslinya berdasarkan pola matrix transpose
    rot = bytearray(BLOCK_SIZE)
    for i in range(BLOCK_SIZE):
        rot[TRANSPOSE_PATTERN[i]] = block[i]

    # Undo Step 2 - Geser ke kanan secara sirkular sebanyak 3 posisi
    shift = 3
    sub = bytearray(BLOCK_SIZE)
    for i in range(BLOCK_SIZE):
        sub[(i + shift) % BLOCK_SIZE] = rot[i]

    # Undo Step 1 - Membalikkan substitusi: byte - key_byte (mod 256)
    out = bytearray(BLOCK_SIZE)
    for i in range(BLOCK_SIZE):
        out[i] = (sub[i] - key[i] + 256) % 256

    return bytes(out)


def encrypt_ecb(data: bytes, key_str: str, use_padding: bool) -> bytes:
    key = pad_key(key_str)
    process_data = block_pad(data) if use_padding else data
    num_blocks = len(process_data) // BLOCK_SIZE

    out = bytearray(process_data)
    for i in range(num_blocks):
        offset = i * BLOCK_SIZE
        block       = process_data[offset : offset + BLOCK_SIZE]
        cipher_block = _encrypt_block_ecb(block, key)
        out[offset : offset + BLOCK_SIZE] = cipher_block

    return bytes(out)


def decrypt_ecb(data: bytes, key_str: str, use_padding: bool) -> bytes:
    key = pad_key(key_str)
    num_blocks = len(data) // BLOCK_SIZE

    out = bytearray(data)
    for i in range(num_blocks):
        offset = i * BLOCK_SIZE
        block       = data[offset : offset + BLOCK_SIZE]
        plain_block = _decrypt_block_ecb(block, key)
        out[offset : offset + BLOCK_SIZE] = plain_block

    result = bytes(out)
    return block_unpad(result) if use_padding else result


# ─────────────────────────────────────────────
#  CBC - Cipher Block Chaining Mode
#  XOR-sub -> Rotate-Left-5 -> Pairwise-swap+NOT
# ─────────────────────────────────────────────

def _encrypt_block_cbc(block: bytes, key: bytes) -> bytes:
    block = bytearray(block)
    key   = bytearray(key)

    # Step 1 - Substitution: XOR dengan key
    sub = bytearray(BLOCK_SIZE)
    for i in range(BLOCK_SIZE):
        sub[i] = block[i] ^ key[i]

    # Step 2 - Rotasi: geser ke kiri secara sirkular sebanyak 5 posisi
    shift = 5
    rot = bytearray(BLOCK_SIZE)
    for i in range(BLOCK_SIZE):
        rot[i] = sub[(i + shift) % BLOCK_SIZE]

    # Step 3 - Permutation: pairwise swap + bitwise NOT
    out = bytearray(BLOCK_SIZE)
    for i in range(0, BLOCK_SIZE, 2):
        out[i]     = (~rot[i + 1]) & 0xFF
        out[i + 1] = (~rot[i])     & 0xFF

    return bytes(out)


def _decrypt_block_cbc(block: bytes, key: bytes) -> bytes:
    block = bytearray(block)
    key   = bytearray(key)

    # Undo Step 3 - Membalikkan Permutation (swap+NOT is self-inverse, jadi kode sama)
    rot = bytearray(BLOCK_SIZE)
    for i in range(0, BLOCK_SIZE, 2):
        rot[i]     = (~block[i + 1]) & 0xFF
        rot[i + 1] = (~block[i])     & 0xFF

    # Undo Step 2 - Geser ke kanan secara sirkular sebanyak 5 posisi
    shift = 5
    sub = bytearray(BLOCK_SIZE)
    for i in range(BLOCK_SIZE):
        sub[(i + shift) % BLOCK_SIZE] = rot[i]

    # Undo Step 1 - XOR dengan key
    out = bytearray(BLOCK_SIZE)
    for i in range(BLOCK_SIZE):
        out[i] = sub[i] ^ key[i]

    return bytes(out)


def encrypt_cbc(data: bytes, key_str: str, iv_str: str, use_padding: bool) -> bytes:
    key = pad_key(key_str)
    iv  = pad_key(iv_str)

    process_data = block_pad(data) if use_padding else data
    num_blocks = len(process_data) // BLOCK_SIZE

    out = bytearray(process_data)
    prev_cipher_block = iv

    for i in range(num_blocks):
        offset = i * BLOCK_SIZE
        block  = process_data[offset : offset + BLOCK_SIZE]

        # CBC chain: XOR plaintext dengan block sebelumnya atau IV
        chained_block = xor_blocks(block, prev_cipher_block)
        cipher_block  = _encrypt_block_cbc(chained_block, key)
        out[offset : offset + BLOCK_SIZE] = cipher_block

        prev_cipher_block = cipher_block  # block selanjutnya didasari oleh blok sebelumnya

    return bytes(out)


def decrypt_cbc(data: bytes, key_str: str, iv_str: str, use_padding: bool) -> bytes:
    key = pad_key(key_str)
    iv  = pad_key(iv_str)

    num_blocks = len(data) // BLOCK_SIZE
    out = bytearray(data)
    prev_cipher_block = iv

    for i in range(num_blocks):
        offset       = i * BLOCK_SIZE
        cipher_block = data[offset : offset + BLOCK_SIZE]

        # Decrypt dulu, XOR kemudian
        decrypted_block = _decrypt_block_cbc(cipher_block, key)
        plain_block     = xor_blocks(decrypted_block, prev_cipher_block)
        out[offset : offset + BLOCK_SIZE] = plain_block

        #gunakan blok cipher asli untuk chaining (bukan yang sudah didecrypt)
        prev_cipher_block = cipher_block

    result = bytes(out)
    return block_unpad(result) if use_padding else result


# ─────────────────────────────────────────────
#  Quick Demo / Self-Test
# ─────────────────────────────────────────────

if __name__ == "__main__":
    key = "secret"
    iv  = "initialverctor"
    plaintext = "101032300039 - Abdu Fattah"

    print("=" * 60)
    print("  BLOCK CIPHER - Python Demo")
    print("=" * 60)
    print(f"  Plaintext : {plaintext!r}")
    print(f"  Key       : {key!r}")

    # ── ECB ──────────────────────────────────
    print("\n[ ECB MODE ]")
    ecb_cipher = encrypt_ecb(plaintext.encode("utf-8"), key, use_padding=True)
    ecb_hex    = bytes_to_hex(ecb_cipher)
    print(f"  Encrypted (hex) : {ecb_hex}")

    ecb_plain  = decrypt_ecb(ecb_cipher, key, use_padding=True)
    print(f"  Decrypted       : {ecb_plain.decode('utf-8')!r}")
    assert ecb_plain.decode("utf-8") == plaintext, "ECB round-trip FAILED!"
    print("  [OK] ECB round-trip passed")

    # ── CBC ──────────────────────────────────
    print("\n[ CBC MODE ]")
    cbc_cipher = encrypt_cbc(plaintext.encode("utf-8"), key, iv, use_padding=True)
    cbc_hex    = bytes_to_hex(cbc_cipher)
    print(f"  Encrypted (hex) : {cbc_hex}")

    cbc_plain  = decrypt_cbc(cbc_cipher, key, iv, use_padding=True)
    print(f"  Decrypted       : {cbc_plain.decode('utf-8')!r}")
    assert cbc_plain.decode("utf-8") == plaintext, "CBC round-trip FAILED!"
    print("  [OK] CBC round-trip passed")

    # ── Hex round-trip ──────────────────────
    print("\n[ HEX HELPERS ]")
    recovered = hex_to_bytes(ecb_hex)
    assert recovered == ecb_cipher
    print("  [OK] bytes_to_hex / hex_to_bytes passed")

    print("\n  All tests passed!")
